actionview/lib/action_view/helpers/controller_helper.rb

```diff
@@ -34,6 +34,7 @@ def logger
       end
 
       def respond_to?(method_name, include_private = false)
+        return true if super.respond_to?(method_name)
```

Comment:
This doesn't seem like the right change to me.

This file lists `cookies` explicitly as something that we should be delegating to the controller. So I'm actually not sure why we are calling `helper_method` at all because I _think_ the delegation here should be making `cookies` available to views already?

Reply:
Thank you for the reply.

Would it be a good idea to explicitly return true for the cookies method since it is already defined as a helper_method?

Reply:
Yeah sorry I think I was unclear in the issue. What I meant that one possible solution is to avoid delegating the check to the controller if `self` already responds to the given method name.
So I was imagining something like
```ruby
      def respond_to?(method_name, include_private = false)
        super_respond_to = super
        return true if super_respond_to
        return controller.respond_to?(method_name) if CONTROLLER_DELEGATES.include?(method_name.to_sym)

       super_respond_to
      end
```

But now I think maybe removing `cookies` from delegation list might be a better option
Though we need to make sure nothing expects the delegation to exist when we don't define the `helper_method :cookies` (Rails API context I presume) but then I don't think Rails API mode is interested in any `cookies` method regardless of where its defined

Reply:
No, I don't think we should be adding another layer to this at all, I think we need to fix the current logic by either:

- removing `helper_method` and let Action View delegate to the cookies method on the controller (since `cookies` is in the list of methods to delegate here)
- or removing `cookies` from the delegation since `helper_method` should be defining `cookies` for Action View

Reply:
>and let Action View delegate to the cookies method on the controller (since cookies is in the list of methods to delegate here)

I need to double check it but I believe this won't work since `cookies` on the controller is `private` and currently the method is only accessible due to `helper_method` making it public. But `controller.cookies` fails

Reply:
@nvasilevski @skipkayhil To make sure I'm correctly understanding things, would the correct solution be to remove cookies from the delegation list?

Reply:
I think removing `:cookies` from the list is the right choice. Unless I'm missing something delegating `cookies` to the controller won't work since it's a private method of the controller and currently `cookies` are accessible in the view due to `helper_method`. 

I wish there was a test showing the fix. I couldn't come up with a quick one but perhaps we should have an integration-level test with a real view that shows that `cookies` method is available from within the view but `respond_to?(:cookies)` returns `false` (which by implication leads to a wrong `defined?` result)

Reply:
@nvasilevski I was also wondering about how to test this fix. I agree; we probably would need an integration test since more than one module may be involved.

Though I do have a question: how would we be able to show that cookies method is available from within a view?

Reply:
check out the test I added in #49858, you should be able to do something similar:

```
def action_name
  cookies[:something] = "whatever"
  render inline: "<%= cookies[:something] %>"
end
```

and then make an assertion on the response

---

to the point about `cookies` being private... it also seems weird to me that we have a private method that we then expose because we call `helper_method` on it. I wonder if removing `helper_method` and just making the method public would be better?